client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java

```diff
@@ -0,0 +1,306 @@
+        return CompletableFuture.supplyAsync(() -> {
+            Socket socket = createClientSocket(location, clientConfig); 
+            try {
+                InputStream inputStream = socket.getInputStream();
+                AppendBatchSizeTrackerImpl batchSizeTracker = new AppendBatchSizeTrackerImpl();
+                ConnectionReader reader = new ConnectionReader(location.toString(), inputStream, callback, batchSizeTracker);
+                reader.start();
+                CommandEncoder encoder = new CommandEncoder(l -> batchSizeTracker, null, socket.getOutputStream());
+                return new TcpClientConnection(socket, encoder, reader, location, onClose, executor);
+            } catch (IOException e) {
+                try {
+                    onClose.run();
+                    socket.close();
+                } catch (IOException e1) {
+                    log.warn("Failed to close socket while failing.", e1);
+                }
+                throw Exceptions.sneakyThrow(e);
+            }
+        }, executor);
+    }
+
+    private static TrustManagerFactory createFromCert(String trustStoreFilePath)
+            throws CertificateException, IOException, NoSuchAlgorithmException, KeyStoreException {
+        TrustManagerFactory factory = null;
+        if (!Strings.isNullOrEmpty(trustStoreFilePath)) {
+            KeyStore trustStore = CertificateUtils.createTrustStore(trustStoreFilePath);
+
+            factory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+            factory.init(trustStore);
+        }
+        return factory;
+    }
+
+    /**
+     * Creates a socket connected to the provided endpoint. 
+     * Note that this is a sync call even though it is called in an async context. 
+     * While this is normally frowned upon, it is simply not possible to construct an SSL socket asynchronously in Java.
+     */
+    @SneakyThrows //Is called inside a completable future.
+    private static Socket createClientSocket(PravegaNodeUri location, ClientConfig clientConfig) {        
+        Socket result;
+        if (clientConfig.isEnableTlsToSegmentStore()) {
+            TrustManagerFactory trustMgrFactory = createFromCert(clientConfig.getTrustStore());
+
+            // Prepare a TLS context that uses the trust manager
+            SSLContext tlsContext = SSLContext.getInstance("TLS");
+            tlsContext.init(null,
+                    trustMgrFactory != null ? trustMgrFactory.getTrustManagers() : null,
+                    null);
+
+            SSLSocket s = (SSLSocket) tlsContext.getSocketFactory().createSocket();
+            // SSLSocket does not perform hostname verification by default. So, we must explicitly enable it.
+            if (clientConfig.isValidateHostName()) {
+                SSLParameters tlsParams = new SSLParameters();
+                tlsParams.setEndpointIdentificationAlgorithm("HTTPS");
```

Comment:
HTTPS ?

Reply:
Yeah, it does look strange as it is connecting to TCP service exposed by Segment Store, and not to an HTTP (application-layer) server. 

However, note that SSLSocket does not perform TLS hostname verification by default and this is a common way of turning it on regardless of the application protocol. We were using the same mechanism elsewhere too: 

https://github.com/pravega/pravega/blob/c7ac009970787df2633163644691aa3d72fc1a4e/client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java#L249

So, this code is alright.

Reply:
Thanks for the explanation.
What about adding a little comment?
This way who will read this code won't be worried

Reply:
I've added the comment in the latest commit I just pushed.